src/components/SystemPreferences.vue

```diff
@@ -105,23 +105,44 @@ export default {
         ],
       ];
     },
-    updatedMemory(value) {
+    updatedVal(value, key = 'memory') {
+      const unit = key === 'memory' ? 'GB' : 'CPUs';
       let warningMessage = '';
 
-      if (value > this.availMemoryInGB - this.safeReservedMemoryInGB) {
-        warningMessage = `Allocating ${ value } GB to the virtual machine may cause your host machine to be sluggish.`;
+      if (this.hasError(key, value)) {
+        const comparison = (key === 'memory' && value > this.availMemoryInGB) || (key === 'cpu' && value > this.availNumCPUs) ? 'Less' : 'More';
+        const threshold = this.threshold(key, comparison);
+
+        this.$emit('error', key, `${ comparison } than ${ threshold } ${ unit } needs to be allocated to the virtual machine.`);
+
+        return;
       }
-      this.$emit('warning', 'memory', warningMessage);
-      this.$emit('updateMemory', value);
-    },
-    updatedCPU(value) {
-      let warningMessage = '';
 
-      if (value > this.availNumCPUs - this.reservedNumCPUs) {
-        warningMessage = `Allocating ${ value } CPUs to the virtual machine may cause your host machine to be sluggish.`;
+      if (this.hasWarning(key, value)) {
+        warningMessage = `Allocating ${ value } ${ unit } to the virtual machine may cause your host machine to be sluggish.`;
+      }
+
+      this.$emit('warning', key, warningMessage);
+      this.$emit(`update:${ key }`, Number(value));
+    },
+    hasError(key, val) {
+      return (
```

Comment:
So part of this PR is to replace specific functions with more generic ones, which works fine for `updatedVal`. I was curious about doing something like:
```
hasError(key, val) {
  cons func = {memory: () => (val > this.availMemoryInGB || val < this.safeMinMemory),
                       cpu: () => (val > this.availNumCPUs || val < this.safeMinCPUs)}[key];
  return func && func();
  }
```
  
  but wrote a quick benchmark and it's about 40 times slower, so don't do that :)

Reply:
Thanks for the heads up, I like where you're head is at! I was trying to distill this down in a not-so-similar fashion, but readability took a nosedive real quick. The boring approach became incredibly enticing.. 😁